pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartialUpsertHandler.java

```diff
@@ -0,0 +1,141 @@
+      return true;
+    }
+
+    HelixDataAccessor dataAccessor = _helixManager.getHelixDataAccessor();
+    PropertyKey.Builder keyBuilder = dataAccessor.keyBuilder();
+    IdealState idealState = dataAccessor.getProperty(keyBuilder.idealStates(_tableNameWithType));
+    if (idealState == null) {
+      LOGGER.warn("Failed to find ideal state for table: {}", _tableNameWithType);
+      return false;
+    }
+    String instanceName = _helixManager.getInstanceName();
+    LiveInstance liveInstance = dataAccessor.getProperty(keyBuilder.liveInstance(instanceName));
+    if (liveInstance == null) {
+      LOGGER.warn("Failed to find live instance for instance: {}", instanceName);
+      return false;
+    }
+    String sessionId = liveInstance.getEphemeralOwner();
+    CurrentState currentState =
+        dataAccessor.getProperty(keyBuilder.currentState(instanceName, sessionId, _tableNameWithType));
+    if (currentState == null) {
+      LOGGER.warn("Failed to find current state for instance: {}, sessionId: {}, table: {}", instanceName, sessionId,
+          _tableNameWithType);
+      return false;
+    }
+
+    // Check if ideal state and current state matches for all segments assigned to the current instance
+    Map<String, Map<String, String>> idealStatesMap = idealState.getRecord().getMapFields();
+    Map<String, String> currentStateMap = currentState.getPartitionStateMap();
+    for (Map.Entry<String, Map<String, String>> entry : idealStatesMap.entrySet()) {
+      String segmentName = entry.getKey();
+      Map<String, String> instanceStateMap = entry.getValue();
+      String expectedState = instanceStateMap.get(instanceName);
+      // Only track ONLINE segments assigned to the current instance
+      if (!SegmentStateModel.ONLINE.equals(expectedState)) {
+        continue;
+      }
+      String actualState = currentStateMap.get(segmentName);
+      if (!SegmentStateModel.ONLINE.equals(actualState)) {
+        if (SegmentStateModel.ERROR.equals(actualState)) {
+          LOGGER
+              .error("Find ERROR segment: {}, table: {}, expected: {}", segmentName, _tableNameWithType, expectedState);
+        } else {
+          LOGGER.info("Find unloaded segment: {}, table: {}, expected: {}, actual: {}", segmentName, _tableNameWithType,
+              expectedState, actualState);
+        }
+        return false;
+      }
+    }
+
+    LOGGER.info("All segments loaded for table: {}", _tableNameWithType);
+    _allSegmentsLoaded = true;
+    return true;
+  }
+
+  /**
+   * Merges 2 records and returns the merged record.
+   *
+   * @param previousRecord the last derived full record during ingestion.
+   * @param newRecord the new consumed record.
+   * @return a new row after merge
+   */
+  public GenericRow merge(GenericRow previousRecord, GenericRow newRecord) {
+    for (Map.Entry<String, PartialUpsertMerger> entry : _mergers.entrySet()) {
+      String column = entry.getKey();
+      if (!previousRecord.isNullValue(column)) {
```

Comment:
should it be part of a merge strategy to deal with Null value handling in two records? Why this logic appear here rather than delegating to each mergeStrategy?

Reply:
@chenboat after discuss with @Jackie-Jiang  and @yupeng9 we finalized to update merge interface from merge(row1, row2) to merge (value1, value2). With this change null value handling is moved to the handler and removed duplicated null value handling code in the mergers.